In [ ]:
# Question 2: NumPy Numerical Analysis

## GDP Growth vs Employment by Economic Sector

In [ ]:
## 1. Introduction

#This section performs numerical analysis using NumPy on the cleaned GDP Growth and Employment by Economic Sector dataset. The analysis focuses on applying NumPy arrays and numerical functions to calculate descriptive statistics, perform array operations, reshape data, and identify numerical patterns within the dataset.

#The cleaned dataset used in this analysis was prepared in Question 1 and is stored in the project's `data/clean` folder.

In [3]:
%pip install pandas numpy

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.3 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [4]:
import pandas as pd
import numpy as np

In [5]:
# Load the cleaned dataset
df = pd.read_csv("../../data/clean/GDP_Employment_Cleaned.csv")

df.head()

,REF_AREA_LABEL,TIME_PERIOD,GDP_Growth,Employment_Thousands,Sector
0,Zambia,2015,2.920375,4532.103027,Sector: Total
1,Zambia,2015,2.920375,3089.903564,Sector: Agriculture
2,Zambia,2015,2.920375,99.403816,Sector: Mining
3,Zambia,2015,2.920375,180.855972,Sector: Manufacturing
4,Zambia,2015,2.920375,10.042576,Sector: Utilities


In [9]:
# Display the dataset dimensions
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

# Display column names
print("Columns:")
print(df.columns.tolist())

# Display data types
df.dtypes

Number of rows: 37700
Number of columns: 5
Columns:
['REF_AREA_LABEL', 'TIME_PERIOD', 'GDP_Growth', 'Employment_Thousands', 'Sector']


REF_AREA_LABEL              str
TIME_PERIOD               int64
GDP_Growth              float64
Employment_Thousands    float64
Sector                      str
dtype: object

In [10]:
# Check for missing values in the cleaned dataset
print("Missing values:")
print(df.isnull().sum())

# Check for duplicate rows
print("\nDuplicate rows:", df.duplicated().sum())

Missing values:
REF_AREA_LABEL          0
TIME_PERIOD             0
GDP_Growth              0
Employment_Thousands    0
Sector                  0
dtype: int64

Duplicate rows: 0


## 2. Numerical Data Preparation

The cleaned dataset contains 37,700 observations and five variables. For the NumPy numerical analysis, the numerical variables selected are `TIME_PERIOD`, `GDP_Growth`, and `Employment_Thousands`. The categorical variables `REF_AREA_LABEL` and `Sector` are used to identify countries and economic sectors but are not directly used for numerical calculations.

## 3. NumPy Array Conversion and Descriptive Statistics

In [11]:
# Convert numerical columns to NumPy arrays

years = df["TIME_PERIOD"].to_numpy()
gdp_growth = df["GDP_Growth"].to_numpy()
employment = df["Employment_Thousands"].to_numpy()

print("Years array shape:", years.shape)
print("GDP Growth array shape:", gdp_growth.shape)
print("Employment array shape:", employment.shape)

Years array shape: (37700,)
GDP Growth array shape: (37700,)
Employment array shape: (37700,)


In [12]:
# Display the first 10 values from each NumPy array

print("Years:", years[:10])
print("GDP Growth:", gdp_growth[:10])
print("Employment:", employment[:10])

Years: [2015 2015 2015 2015 2015 2015 2015 2015 2015 2015]
GDP Growth: [2.92037511 2.92037511 2.92037511 2.92037511 2.92037511 2.92037511
 2.92037511 2.92037511 2.92037511 2.92037511]
Employment: [4532.10302734 3089.90356445   99.40381622  180.85597229   10.04257584
   35.70127106  795.23620605   70.21607208   47.9518013   202.79154968]


In [13]:
# Descriptive statistics for GDP Growth

print("GDP Growth Mean:", np.mean(gdp_growth))
print("GDP Growth Median:", np.median(gdp_growth))
print("GDP Growth Minimum:", np.min(gdp_growth))
print("GDP Growth Maximum:", np.max(gdp_growth))
print("GDP Growth Standard Deviation:", np.std(gdp_growth))
print("GDP Growth Variance:", np.var(gdp_growth))

GDP Growth Mean: 4.04977657846982
GDP Growth Median: 3.9534899458060124
GDP Growth Minimum: -50.24806710352907
GDP Growth Maximum: 35.224078310176
GDP Growth Standard Deviation: 4.433296814305468
GDP Growth Variance: 19.65412064373101


In [14]:
# Descriptive statistics for Employment

print("Employment Mean:", np.mean(employment))
print("Employment Median:", np.median(employment))
print("Employment Minimum:", np.min(employment))
print("Employment Maximum:", np.max(employment))
print("Employment Standard Deviation:", np.std(employment))
print("Employment Variance:", np.var(employment))

Employment Mean: 5222.356916579864
Employment Median: 339.66716733067733
Employment Minimum: 0.065
Employment Maximum: 775772.0625
Employment Standard Deviation: 30992.609509177746
Employment Variance: 960541844.1883748


### Interpretation

The NumPy descriptive statistics provide an overview of the distribution of GDP growth and employment values in the cleaned dataset. The mean and median provide measures of central tendency, while the minimum and maximum values show the range of the observations. Standard deviation and variance indicate the amount of variation in the numerical data.

These statistics provide a numerical foundation for understanding the relationship between economic growth and employment across countries, years, and economic sectors.

## 3. NumPy Array Conversion and Descriptive Statistics

The numerical variables were converted from Pandas Series into NumPy arrays to perform numerical analysis. NumPy was used to calculate measures of central tendency and dispersion, including the mean, median, minimum, maximum, standard deviation, and variance.

The GDP Growth results show a mean of approximately 4.05% and a median of approximately 3.95%. The standard deviation of approximately 4.43 indicates variation in GDP growth across the observations.

Employment shows a much larger difference between its mean and median. The mean employment value is approximately 5,222.36 thousand, while the median is approximately 339.67 thousand. This large difference, together with the maximum value of approximately 775,772.06 thousand, indicates substantial variation in employment values across the observations.

In [15]:
# Calculate additional NumPy statistics

gdp_range = np.ptp(gdp_growth)
employment_range = np.ptp(employment)

gdp_percentiles = np.percentile(gdp_growth, [25, 50, 75])
employment_percentiles = np.percentile(employment, [25, 50, 75])

print("GDP Growth Range:", gdp_range)
print("GDP Growth 25th, 50th and 75th Percentiles:", gdp_percentiles)

print("\nEmployment Range:", employment_range)
print("Employment 25th, 50th and 75th Percentiles:", employment_percentiles)

GDP Growth Range: 85.47214541370506
GDP Growth 25th, 50th and 75th Percentiles: [1.93913775 3.95348995 6.11957879]

Employment Range: 775771.9975
Employment 25th, 50th and 75th Percentiles: [  60.16853619  339.66716733 1985.42605586]


### Interpretation of Range and Percentiles

The range measures the difference between the maximum and minimum values, while the 25th, 50th and 75th percentiles describe the distribution of the observations. Comparing these values helps identify how widely GDP growth and employment values vary across the dataset.

The median corresponds to the 50th percentile and provides a useful comparison with the mean when the data contains extreme values.

In [16]:
# Basic NumPy array operations

gdp_squared = np.square(gdp_growth)
employment_log = np.log1p(employment)

print("Original GDP Growth (first 10):")
print(gdp_growth[:10])

print("\nSquared GDP Growth (first 10):")
print(gdp_squared[:10])

print("\nLog-transformed Employment (first 10):")
print(employment_log[:10])

Original GDP Growth (first 10):
[2.92037511 2.92037511 2.92037511 2.92037511 2.92037511 2.92037511
 2.92037511 2.92037511 2.92037511 2.92037511]

Squared GDP Growth (first 10):
[8.5285908 8.5285908 8.5285908 8.5285908 8.5285908 8.5285908 8.5285908
 8.5285908 8.5285908 8.5285908]

Log-transformed Employment (first 10):
[8.41916198 8.03621874 4.60920022 5.20321501 2.40175833 3.60281139
 6.67989588 4.26571852 3.89083617 5.31709766]


In [17]:
# Identify GDP growth values above and below the mean

gdp_mean = np.mean(gdp_growth)

above_mean = gdp_growth[gdp_growth > gdp_mean]
below_mean = gdp_growth[gdp_growth < gdp_mean]

print("GDP Growth Mean:", gdp_mean)
print("Number of observations above the mean:", above_mean.size)
print("Number of observations below the mean:", below_mean.size)

GDP Growth Mean: 4.04977657846982
Number of observations above the mean: 18450
Number of observations below the mean: 19250


## 5. NumPy Array Reshaping

Reshaping changes the structure of an array without changing the underlying values. This is useful when numerical data needs to be represented in a matrix-like structure for further analysis.

In [18]:
# Reshape GDP Growth into a two-dimensional column array

gdp_matrix = gdp_growth.reshape(-1, 1)

print("Original GDP array shape:", gdp_growth.shape)
print("Reshaped GDP array shape:", gdp_matrix.shape)

print("\nFirst 5 rows of reshaped array:")
print(gdp_matrix[:5])

Original GDP array shape: (37700,)
Reshaped GDP array shape: (37700, 1)

First 5 rows of reshaped array:
[[2.92037511]
 [2.92037511]
 [2.92037511]
 [2.92037511]
 [2.92037511]]


## 6. Relationship Between GDP Growth and Employment

The dataset contains GDP growth and employment observations for different countries, years, and economic sectors. NumPy can be used to calculate the numerical relationship between these two variables. The Pearson correlation coefficient is used to determine whether GDP growth and employment tend to move together in the dataset.

In [19]:
# Calculate the Pearson correlation coefficient using NumPy

correlation_matrix = np.corrcoef(gdp_growth, employment)
correlation = correlation_matrix[0, 1]

print("GDP Growth vs Employment correlation:", correlation)

GDP Growth vs Employment correlation: 0.06357931366610184


In [20]:
# Classify GDP growth observations using NumPy Boolean indexing

positive_gdp = gdp_growth[gdp_growth > 0]
negative_gdp = gdp_growth[gdp_growth < 0]
zero_gdp = gdp_growth[gdp_growth == 0]

print("Positive GDP growth observations:", positive_gdp.size)
print("Negative GDP growth observations:", negative_gdp.size)
print("Zero GDP growth observations:", zero_gdp.size)

print("\nMean positive GDP growth:", np.mean(positive_gdp))
print("Mean negative GDP growth:", np.mean(negative_gdp))

Positive GDP growth observations: 33510
Negative GDP growth observations: 4190
Zero GDP growth observations: 0

Mean positive GDP growth: 4.961709484541451
Mean negative GDP growth: -3.2435102192534133


In [22]:
# Create a numerical summary of employment by sector

sector_names = []
sector_mean_employment = []

for sector in sectors:
    sector_mask = df["Sector"].to_numpy() == sector
    sector_employment = employment[sector_mask]
    
    sector_names.append(sector)
    sector_mean_employment.append(np.mean(sector_employment))

sector_mean_employment = np.array(sector_mean_employment)

for i in range(len(sector_names)):
    print(f"{sector_names[i]}: {sector_mean_employment[i]:,.2f} thousand")

Sector: Total: 26,111.78 thousand
Sector: Agriculture: 10,765.85 thousand
Sector: Mining: 235.39 thousand
Sector: Manufacturing: 3,719.40 thousand
Sector: Utilities: 161.45 thousand
Sector: Construction: 1,497.86 thousand
Sector: Trade services: 3,299.96 thousand
Sector: Transport services: 1,216.77 thousand
Sector: Finance and business services: 1,007.06 thousand
Sector: Other services: 4,208.04 thousand


## 9. Standardised Numerical Analysis

Standardisation allows numerical observations to be compared relative to their mean and standard deviation. A z-score indicates how many standard deviations an observation is above or below the mean.

In [25]:
# Calculate z-scores for GDP Growth and Employment

gdp_z_scores = (gdp_growth - np.mean(gdp_growth)) / np.std(gdp_growth)
employment_z_scores = (employment - np.mean(employment)) / np.std(employment)

print("GDP Growth z-scores (first 10):")
print(gdp_z_scores[:10])

print("\nEmployment z-scores (first 10):")
print(employment_z_scores[:10])

GDP Growth z-scores (first 10):
[-0.25475431 -0.25475431 -0.25475431 -0.25475431 -0.25475431 -0.25475431
 -0.25475431 -0.25475431 -0.25475431 -0.25475431]

Employment z-scores (first 10):
[-0.02227156 -0.06880522 -0.16529596 -0.16266784 -0.16817927 -0.16735137
 -0.1428444  -0.16623772 -0.1669561  -0.16196008]


In [29]:
# Identify observations more than two standard deviations from the GDP mean

extreme_gdp = np.abs(gdp_z_scores) > 2

print("GDP observations more than 2 standard deviations from the mean:",
      np.sum(extreme_gdp))

print("Percentage of extreme GDP observations:",
      np.mean(extreme_gdp) * 100, "%")

GDP observations more than 2 standard deviations from the mean: 1750
Percentage of extreme GDP observations: 4.641909814323608 %


In [30]:
# Locate the minimum and maximum GDP growth observations

min_index = np.argmin(gdp_growth)
max_index = np.argmax(gdp_growth)

print("Minimum GDP Growth:", gdp_growth[min_index])
print("Country:", df.iloc[min_index]["REF_AREA_LABEL"])
print("Year:", df.iloc[min_index]["TIME_PERIOD"])
print("Sector:", df.iloc[min_index]["Sector"])

print("\nMaximum GDP Growth:", gdp_growth[max_index])
print("Country:", df.iloc[max_index]["REF_AREA_LABEL"])
print("Year:", df.iloc[max_index]["TIME_PERIOD"])
print("Sector:", df.iloc[max_index]["Sector"])

Minimum GDP Growth: -50.24806710352907
Country: Rwanda
Year: 1994
Sector: Sector: Total

Maximum GDP Growth: 35.224078310176
Country: Rwanda
Year: 1995
Sector: Sector: Total


In [31]:
# Locate the maximum employment observation

max_employment_index = np.argmax(employment)

print("Maximum Employment:", employment[max_employment_index])
print("Country:", df.iloc[max_employment_index]["REF_AREA_LABEL"])
print("Year:", df.iloc[max_employment_index]["TIME_PERIOD"])
print("Sector:", df.iloc[max_employment_index]["Sector"])

Maximum Employment: 775772.0625
Country: China
Year: 2017
Sector: Sector: Total
